pkg/block/fetcher.go

```diff
@@ -716,8 +694,8 @@ func NewReplicaLabelRemover(logger log.Logger, replicaLabels []string) *ReplicaL
 	return &ReplicaLabelRemover{logger: logger, replicaLabels: replicaLabels}
 }
 
-// Modify modifies external labels of existing blocks, it removes given replica labels from the metadata of blocks that have it.
-func (r *ReplicaLabelRemover) Modify(_ context.Context, metas map[ulid.ULID]*metadata.Meta, modified *extprom.TxGaugeVec) error {
+// Filter modifies external labels of existing blocks, it removes given replica labels from the metadata of blocks that have it.
+func (r *ReplicaLabelRemover) Filter(_ context.Context, metas map[ulid.ULID]*metadata.Meta, synced *extprom.TxGaugeVec, modified *extprom.TxGaugeVec) error {
```

Comment:
This is the most interesting part of this PR :thinking: trying to understand this. Why can't we leave the old `Modify` interface and apply it to each group individually?

Reply:
The concrete problem is sequencing. The replica label modifier changes what groups things are in so it needs to happen before deduplication or you actually do the wrong thing in the presence of replica labels. And you can't just move modifiers to be first because relabeling needs to see the original labels. So it has to fit into the middle of the filter list between those steps.

The two interfaces were identical other than the name of the metric to update so it seemed both cleanest and easiest to just pass both metrics to all of them and merge them into one pipeline of processing stages.

Reply:
LGTM to me